ale_linters/sml/smlnj.vim

```diff
@@ -39,9 +45,33 @@ function! ale_linters#sml#smlnj#Handle(buffer, lines) abort
     return l:out
 endfunction
 
+function! ale_linters#sml#smlnj#GetCommand(buffer) abort
+  let l:pattern = ale#Var(a:buffer, 'sml_smlnj_cm_file')
+  let l:as_list = 1
+
+  let l:cmfile = ''
+  for l:path in ale#path#Upwards(expand('#' . a:buffer . ':p:h'))
+    let l:results = glob(l:pattern, 0, l:as_list)
+    if len(l:results) > 0
+      " If there is more than one CM file, we take the first one
+      " See :help ale-sml-smlnj for how to configure this.
+      let l:cmfile = l:results[0]
+    endif
+  endfor
+
+  if l:cmfile ==# ''
+    " No CM file found; default to checking just this file
+    return 'sml %s < /dev/null'
+  else
+    " Found a CM file; let's use it
+    return 'sml -m ' . l:cmfile . ' < /dev/null'
+  endif
+endfunction
+
 call ale#linter#Define('sml', {
 \   'name': 'smlnj',
 \   'executable': 'sml',
-\   'command': 'sml',
+\   'lint_file': 1,
```

Comment:
We're losing the ability to check for problems while you type here. Is there some way we can keep this behaviour, or is there some technical reason for why we have to check the file on disk?

Reply:
Is there a way to set `lint_file` dynamically from the `GetCommand` function? Or is it the kind of thing that once it's set in `Define` it's unchangable?

The way we invoke the project-level build tool is with `sml -m sources.cm`. The `sources.cm` file is going to declare all the dependencies for a file; CM (Compilation Manager) then finds all the files on disk and compiles the entire project. This is the only way to correctly resolve symbols that are defined in different files.

Unlike other languages, SML doesn't have "import" statements that say one files depends on another file. You declare all files that are used together up front in a list in the CM file, and CM computes the dependency graph for your project.

The way it works for a single file, SML need to use CM. Instead, it just pretends to interpret that one file; if you're interpreting a file which depends on external symbols, they're just going to be not found.

If there is a way to say "in single file mode, use `lint_file: 0`, otherwise `lint_file: 1`, I can add that.

Reply:
I see. I had considered adding something like `lint_file_callback` for dynamically computing the value. In this case, it seems like what you're doing here is just the right thing to do.

We could consider creating two different linters here which run the same program. If `sml` supports options similar to the options `gcc` has for only checking syntax, we could consider writing one linter which only checks for syntax errors in a buffer as you type, and another linter which looks for semantic errors when you open or save files, etc. That might be our best option.

Reply:
If we can't disable semantic analysis, we could also consider filtering out semantic errors when running `sml` against a buffer.